.github/workflows/python-app.yml

```
@@ -28,3 +27,26 @@ jobs:
       run: docker build --target test_e2e --tag todo_app:test_e2e .
     - name: Run e2e tests
       run: docker run --env-file .env todo_app:test_e2e
+
+  deploy:
+    name: Deploy application
+    runs-on: ubuntu-latest
+    needs: [build_and_test]
+    if: ${{ (github.event_name == 'push') && (github.ref == 'refs/heads/master') }}
+    steps:
+    - uses: actions/checkout@v2
+    - name: Login to Docker
+      run: docker login --username ${{ secrets.DOCKER_USERNAME }} --password ${{ secrets.DOCKER_PASSWORD }}
+    - name: Build and push to Docker Hub
+      run: |
+        docker build --target production --tag ${{ secrets.DOCKER_USERNAME }}/todo_app:latest .
+        docker push ${{ secrets.DOCKER_USERNAME }}/todo_app:latest
+    - name: Rename tag to Heroku app name
+      run: docker tag ${{ secrets.DOCKER_USERNAME }}/todo_app registry.heroku.com/${{ secrets.HEROKU_APP_NAME }}/web
+    - name: Login to docker registry
+      run: docker login --username ${{ secrets.DOCKER_USERNAME }} --password ${{ secrets.HEROKU_API_KEY }} registry.heroku.com
+    - name: Push to heroku registry
+      run: docker push registry.heroku.com/${{ secrets.HEROKU_APP_NAME }}/web
+    - name: Release app
+      run: HEROKU_API_KEY=${{ secrets.HEROKU_API_KEY }} heroku container:release web -a ${{ secrets.HEROKU_APP_NAME }}
```

More preference than particularly strong advice, but I tend to find these easier to read without all the secret syntax, so I'd advise putting an env block on the job that turns secrets into environment variables, something like:
```yml
    env:
        HEROKU_API_KEY: ${{ secrets.HEROKU_API_KEY }}
        HEROKU_APP_NAME: ${{ secrets.HEROKU_APP_NAME}}
        ....
```

And you can then refer to them a little easier (and for e.g. the HEROKU_API_KEY we no longer need to set it again here)
```suggestion
      run: heroku container:release web -a $HEROKU_APP_NAME
```